walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,78 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    }
+    ArrayList<TimeRange> occupiedSlotsMandatory =  new ArrayList<TimeRange>();
+    ArrayList<TimeRange> occupiedSlotsAll =  new ArrayList<TimeRange>();
+    boolean foundAnyAttendee;
+    for (Event event: events) {
+        foundAnyAttendee = false;
+        for (String attendee: event.getAttendees()) {
+            if (request.getAttendees().contains(attendee)) {
+                occupiedSlotsMandatory.add(event.getWhen());
+                occupiedSlotsAll.add(event.getWhen());
+                break;
+            }
+            if (foundAnyAttendee == false && request.getOptionalAttendees().contains(attendee)) {
+                occupiedSlotsAll.add(event.getWhen());
+                foundAnyAttendee = true;
+            }
+        }
+    }
+    ArrayList<TimeRange> SlotsForAllAttendees = getOptionalSlots(occupiedSlotsAll, request);
+    if (SlotsForAllAttendees.isEmpty()) {
+        return getOptionalSlots(occupiedSlotsMandatory, request);
+    }  
+    return SlotsForAllAttendees;
   }
+
+    public ArrayList<TimeRange> getOptionalSlots(ArrayList<TimeRange> occupiedSlots, MeetingRequest request) {
```

Comment:
Return List<TimeRange> (not ArrayList), and change the parameter type in the query method to match.

Reply:
Why is List better? because it's more general?